# (공부) Class(클래스)_심화
> 작성 완료

- toc:true
- branch: master
- badges: true
- comments: true
- author: kimha02
- categories: [Python]

## **클래스, 인스턴스, 오브젝트**

`-` 오브젝트
- 클래스 오브젝트
- 인스턴스 오브젝트 

`-` 클래스 (=클래스 오브젝트)

`-` 인스턴스 (=인스턴스 오브젝트)

----

## 클래스 속성 vs 인스턴스 속성

- 노트(5)에서 아래와 같은 노트가 있었다.

> 규칙2: 클래스 내에서 정의한 변수 (예를들면 `title`, `img`, `don`)를 사용하려면
    - `self.title`, `self.img`, `self.don`
    - `MooYaHo.title`, `MooYaHo.img`, `MooYaHo.don`

$\to$  `self.`는 인스턴스 속성, `MooYaHo.`는 클래스 속성을 의미한다.

### [예제1]

In [ ]:
class Testclass1: 
    x=0 
    y=0
    def my_print(self):
        self.x += 1 
        Testclass1.y +=1 
        print("현재 인스턴스에서 %s 회 출력" % self.x)
        print("전체 인스턴스에서 총 %s 회 출력" % self.y)

In [ ]:
f=Testclass1

In [ ]:
a=Testclass1()

In [ ]:
b=f()  ### f라는 클래스 오브젝트에서 b라는 인스턴트 오브젝트 생성

In [ ]:
a.my_print()

In [ ]:
b.my_print()

In [ ]:
b.my_print()

In [ ]:
a.my_print()

In [ ]:
a.my_print()

`-` 신기한 점: 각 인스턴스에서 instance.my_print()를 실행한 횟수를 서로 공유하는 듯 하다. 

#### 분석

`-` 코드를 시점별로 분석해보자. 

`-` 분석을 위해서 커널을 재시작한다. 

[시점1]: Testclass1를 선언하는 시점

In [1]:
class Testclass1: 
    x=0 
    y=0
    def my_print(self):
        self.x += 1 
        Testclass1.y +=1 
        print("현재 인스턴스에서 %s 회 출력" % self.x)
        print("전체 인스턴스에서 총 %s 회 출력" % self.y)

In [2]:
dir(Testclass1)   ###Testclass 탐색

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 'my_print',
 'x',
 'y']

In [3]:
dir(a)

NameError: name 'a' is not defined

In [4]:
dir(b)

NameError: name 'b' is not defined

`–` 이 시점에는 Testclass1만이 존재한다. Testclass1를 바로 `클래스 오브젝트` 라고 부름. 

In [5]:
Testclass1.x

0

In [6]:
Testclass1.y

0

`–` 현재시점에서는 클래스 오브젝트의 수 1개, 인스턴스 오브젝트의 수 0개, 따라서 총 오브젝트 수는 1개임. 

[시점2] 클래스에 별칭을 지정하는 시점 

In [8]:
f=Testclass1 ###f라는 별칭을 지정

In [9]:
f.x

0

In [10]:
f.y

0

In [11]:
Testclass1.x

0

In [12]:
Testclass1.y

0

`–` 이 시점에서 클래스 오브젝트는 2개가 있는 것 처럼 보인다. 

`-` 그렇다면 이 2개의 클래스 오브젝트는 컴퓨터의 어딘가에 저장이 되어 있을 것이다. 

`-` 구체적으로는 메모리에 저장되어 있을 것. 

`-` 2개의 클래스오브젝트는 서로 다른 메모리 공간에 저장되어 있을 것이다.

`-` 진짜인가? 확인해보자. `id()`는 오브젝트(클래스 오브젝트, 인스턴스 오브젝트)가 저장된 메모리 주소를 확인하는 명령어이다. 

In [13]:
id(f)

93883369152752

`–` f라는 오브젝트는 93883369152752 메모리에 저장되어 있다. 

In [14]:
id(Testclass1)

93883369152752

`-` Testclass1의 오브젝트 역시 93883369152752 메모리에 저장되어 있다. 

`-` 추론: 사실 93883369152752
라는 메모리공간에 저장된 어떠한 것은 동일한데, 그것을 어떤사람은 `Testclass1` 이라고 부르고 어떤사람은 `f`라고 부른다. 

`-` 이는 마치 **별명**이랑 비슷하다. 부르는 이름이 2개라고 해서 나라는 오브젝트가 2개가 있는것은 아니다. 

`-` 결국 이 시점에서 클래스 오브젝트의 수는 여전히 1개라고 볼 수 있다. (인스턴스 오브젝트의 수는 0개)

[시점3]: 클래스 오브젝트로부터 인스턴스 오브젝트를 만드는 시점

In [15]:
a=Testclass1()
b=f()

In [16]:
id(Testclass1),id(f),id(a),id(b)

(93883369152752, 93883369152752, 140489211063024, 140489211062064)

`–` 이 순간에는 클래스 오브젝트 1개, 인스턴스 오브젝트 2개 존재한다. 즉 총 3개의 오브젝트가 존재한다. 

`-` 메모리주소 93883369152752 에 존재하는 오브젝트는 클래스 오브젝트이며 Testclass1 또는 f 라고 불린다. 

`-` 메모리주소 140489211063024 에 존재하는 오브젝트는 인스턴스 오브젝트이며 a라고 불린다. 

`-` 메모리주소 140489211062064 에 존재하는 오브젝트는 인스턴스 오브젝트이며 b라고 불린다. 

In [17]:
Testclass1.x, Testclass1.y

(0, 0)

In [18]:
f.x,f.y

(0, 0)

In [19]:
a.x,a.y

(0, 0)

In [20]:
b.x,b.y

(0, 0)

[시점4]

In [21]:
a.my_print()

현재 인스턴스에서 1 회 출력
전체 인스턴스에서 총 1 회 출력


In [22]:
(f.x,f.y),(a.x,a.y),(b.x,b.y)

((0, 1), (1, 1), (0, 1))

`-` 특징 
 - a.my_print()를 실행하면 a.x 의 값이 1이 증가한다. 
 - a.my_print()를 실행하면 f.y, a.y, b.y 의 값이 동시에 1이 증가한다. (공유가 되는 느낌)

[시점5]

In [23]:
b.my_print()

현재 인스턴스에서 1 회 출력
전체 인스턴스에서 총 2 회 출력


In [24]:
(f.x,f.y),(a.x,a.y),(b.x,b.y)

((0, 2), (1, 2), (1, 2))

[시점6]

In [25]:
b.my_print()

현재 인스턴스에서 2 회 출력
전체 인스턴스에서 총 3 회 출력


In [26]:
(f.x,f.y),(a.x,a.y),(b.x,b.y)

((0, 3), (1, 3), (2, 3))

[시점7]

In [27]:
a.my_print()

현재 인스턴스에서 2 회 출력
전체 인스턴스에서 총 4 회 출력


In [28]:
(f.x,f.y),(a.x,a.y),(b.x,b.y)

((0, 4), (2, 4), (2, 4))

[시점8]

In [29]:
a.my_print()

현재 인스턴스에서 3 회 출력
전체 인스턴스에서 총 5 회 출력


In [30]:
(f.x,f.y),(a.x,a.y),(b.x,b.y)

((0, 5), (3, 5), (2, 5))

---

### [예제2]

`-` 아래처럼 코드를 바꿔도 잘 동작할것 같다. 

In [33]:
class Testclass2: 
    def __init__(self):
        self.x=0
        self.y=0
    def my_print(self):
        self.x += 1 
        Testclass2.y +=1 
        print("현재 인스턴스에서 %s 회 출력" % self.x)
        print("전체 인스턴스에서 총 %s 회 출력" % self.y)

In [34]:
c=Testclass2()

In [35]:
c.my_print()

AttributeError: type object 'Testclass2' has no attribute 'y'

`–` 왜 에러가 나는가? 

In [36]:
dir(Testclass2)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 'my_print']

In [37]:
dir(Testclass1)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 'my_print',
 'x',
 'y']

`-` 관찰1: Testclass2에서는 Testclass1과는 다르게 x,y가 없다. 

In [38]:
dir(c)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 'my_print',
 'x',
 'y']

`–` 관찰2: 그런데 `c`라는 인스턴스 오브젝트에서는 x,y가 있다. 

`-` 추론: `__init__`함수는 클래스 오브젝트가 만들어지는 시점에서는 실행되지 않고, **인스텐스 오브젝트가 만들어지는 시점에 실행된다.**

`-` 결국 `__init__`함수의 역할은 클래스 오브젝트에서 인스턴스 오브젝트를 만든후에 초기화를 위해서 실행하는 어떠한 일련의 명령들을 묶어놓은 것에 불과하다. 

`–` 즉 위의 코드는 굳이 따지면 아래를 실행한 것과 동일하다. 

In [39]:
class Testclass2: 
#     def __init__(self):
#         self.x=0
#         self.y=0
    def my_print(self):
        self.x += 1 
        Testclass2.y +=1 
        print("현재 인스턴스에서 %s 회 출력" % self.x)
        print("전체 인스턴스에서 총 %s 회 출력" % self.y)

In [40]:
c=Testclass2()

In [41]:
c.x=0
c.y=0

`-` 이 상황에서 
```python
c.my_print()
```
를 실행하면
```python
c.x += 1 
Testclass2.y +=1 
print("현재 인스턴스에서 %s 회 출력" % c.x)
print("전체 인스턴스에서 총 %s 회 출력" % c.y)
```
이 실행되는데, 이때 Testclass2.y 이 정의되어 있지 않으므로 
```python
Testclass2.y +=1 
```
에서 에러가 난다. 

----

### [예제 3]

`-` 그렇다면 아래와 같이 수정하면 어떨까? 

In [43]:
class Testclass3: 
    def __init__(self):
        self.x=0
        Testclass3.y=0
    def my_print(self):
        self.x += 1 
        Testclass3.y +=1 
        print("현재 인스턴스에서 %s 회 출력" % self.x)
        print("전체 인스턴스에서 총 %s 회 출력" % self.y)

In [44]:
a=Testclass3()
b=Testclass3()

In [45]:
a.my_print()

현재 인스턴스에서 1 회 출력
전체 인스턴스에서 총 1 회 출력


In [46]:
b.my_print()

현재 인스턴스에서 1 회 출력
전체 인스턴스에서 총 2 회 출력


In [47]:
a.my_print()

현재 인스턴스에서 2 회 출력
전체 인스턴스에서 총 3 회 출력


In [48]:
a.my_print()

현재 인스턴스에서 3 회 출력
전체 인스턴스에서 총 4 회 출력


In [49]:
b.my_print()

현재 인스턴스에서 2 회 출력
전체 인스턴스에서 총 5 회 출력


In [50]:
b.my_print()

현재 인스턴스에서 3 회 출력
전체 인스턴스에서 총 6 회 출력


`–` Testclass1과 동일한 기능이 수행되는것 같다. 

`-` 그런데 조금만 생각해보면 엉터리라는 것을 알 수 있다. 아래의 코드를 관찰하여보자. 

In [51]:
class Testclass3: 
    def __init__(self):
        self.x=0
        Testclass3.y=0
    def my_print(self):
        self.x += 1 
        Testclass3.y +=1 
        print("현재 인스턴스에서 %s 회 출력" % self.x)
        print("전체 인스턴스에서 총 %s 회 출력" % self.y)
        
a=Testclass3()
a.my_print()
a.my_print()
b=Testclass3()
b.my_print()

현재 인스턴스에서 1 회 출력
전체 인스턴스에서 총 1 회 출력
현재 인스턴스에서 2 회 출력
전체 인스턴스에서 총 2 회 출력
현재 인스턴스에서 1 회 출력
전체 인스턴스에서 총 1 회 출력


`-` Testclass3는 인스턴스를 생성할때마다 `y=0`이 설정된다. 그래서 
```python
b=Testclass3()
```
이 시점에서 의도하지 않게 '전체 인스턴스에서 출력된 횟수'를 의미하는 `y`가 초기화되었다. 

`-` 코드는 엉터리이지만, Testclass3은 의외로 분석할만한 가치가 있다. 특히 위의 실행결과를 시점별로 Testclass1과 비교해보면 재미있다. 

#### ***Testclass1 & Testclass3 비교***

`-` Testclass1

```python
### Testclass1 
## 시점1: 클래스 오브젝트 생성 
class Testclass1: 
    x=0 
    y=0
    def my_print(self):
        self.x += 1 
        Testclass1.y +=1 
        print("현재 인스턴스에서 %s 회 출력" % self.x)
        print("전체 인스턴스에서 총 %s 회 출력" % self.y)
## 시점2: 인스턴스 오브젝트 a를 생성 
a=Testclass1()
## 시점3: a에서 메소드 실행 
a.my_print()
## 시점4: a에서 메소드를 한번 더 실행 
a.my_print()
## 시점5: 인스턴스 오브젝트 b를 생성
b=Testclass1()
## 시점6: b에서 메소드를 실행
b.my_print()
```

```
현재 인스턴스에서 1 회 출력
전체 인스턴스에서 총 1 회 출력
현재 인스턴스에서 2 회 출력
전체 인스턴스에서 총 2 회 출력
현재 인스턴스에서 1 회 출력
전체 인스턴스에서 총 3 회 출력
```

| |시점1 |시점2 |시점3 | 시점4 | 시점5| 시점6|
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|Testclass1.x|0| 0| 0| 0| 0| 0|
|Testclass1.y|0| 0| 1| 2| 2 | 3|
|a.x| 값없음     | 0| 1| 2| 2| 2|
|a.y| 값없음     | 0| 1| 2| 2| 3|
|b.x| 값없음|     값없음| 값없음| 값없음| 0| 1|
|b.y| 값없음|     값없음| 값없음| 값없음| 2| 3|

`–` Testclass3

```python
#### Testclass3
## 시점1: 클래스 오브젝트 생성 
class Testclass3: 
    def __init__(self):
        self.x=0
        Testclass3.y=0
    def my_print(self):
        self.x += 1 
        Testclass3.y +=1 
        print("현재 인스턴스에서 %s 회 출력" % self.x)
        print("전체 인스턴스에서 총 %s 회 출력" % self.y)
## 시점2: 인스턴스 오브젝트 a를 생성 
a=Testclass3()
## 시점3: a에서 메소드 실행 
a.my_print()
## 시점4: a에서 메소드를 한번 더 실행 
a.my_print()
## 시점5: 인스턴스 오브젝트 b를 생성
b=Testclass3()
## 시점6: b에서 메소드를 실행
b.my_print()
```

```
현재 인스턴스에서 1 회 출력
전체 인스턴스에서 총 1 회 출력
현재 인스턴스에서 2 회 출력
전체 인스턴스에서 총 2 회 출력
현재 인스턴스에서 1 회 출력
전체 인스턴스에서 총 1 회 출력
```

| |시점1 |시점2 |시점3 | 시점4 | 시점5| 시점6|
|:-:|:-:|:-:|:-:|:-:|:-:|:-:|
|Testclass3.x|값없음| 값없음| 값없음| 값없음| 값없음| 값없음|
|Testclass3.y|값없음| 0| 1| 2| 0 | 1|
|a.x| 값없음     | 0| 1| 2| 2| 2|
|a.y| 값없음     | 0| 1| 2| 0| 1|
|b.x| 값없음|     값없음| 값없음| 값없음| 0| 1|
|b.y| 값없음|     값없음| 값없음| 값없음| 0| 1|

`–` Testclass3.y가 업데이트 되면 a.y, b.y도 자동으로 업데이트 된다. 